eth/downloader/downloader.go

```diff
@@ -477,15 +480,22 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
 			return err
 		}
 		if latest.Number.Uint64() > uint64(fsMinFullBlocks) {
-			pivot = d.skeleton.Header(latest.Number.Uint64() - uint64(fsMinFullBlocks))
+			number := latest.Number.Uint64() - uint64(fsMinFullBlocks)
+			pivot = d.skeleton.Header(number)
+			if pivot == nil {
+				// Retrieve the pivot point from the local chain if it's
+				// not in the range of skeleton.
+				pivot = d.lightchain.GetHeaderByNumber(number)
+			}
 		}
 	}
 	// If no pivot block was returned, the head is below the min full block
 	// threshold (i.e. new chain). In that case we won't really snap sync
 	// anyway, but still need a valid pivot block to avoid some code hitting
 	// nil panics on access.
+	var noSnap bool
 	if mode == SnapSync && pivot == nil {
-		pivot = d.blockchain.CurrentBlock().Header()
+		pivot, noSnap = d.blockchain.CurrentBlock().Header(), true
```

Comment:
What is the semantics of `noSnap`? It is by default `false`, but set to `true` if `mode==SnapSync`, which feels very counter-intuitive to me

Reply:
It's a special case handled here, where the pivot is nil. In this case it means the sync target is below 64, we don't expect any state sync at all.

And this flag is used in Line 532. In line 532, we try to cap the origin with pivot header to ensure the pivot block is synced from the network. However if we don't expect any state sync, this capping is meaningless.

Maybe I should rename this flag?